pkg/skaffold/build/cache/hash.go

```diff
@@ -38,15 +38,72 @@ import (
 
 // For testing
 var (
-	hashFunction           = cacheHasher
-	artifactConfigFunction = artifactConfig
+	newArtifactHasherFunc = newArtifactHasher
+	fileHasherFunc        = fileHasher
+	artifactConfigFunc    = artifactConfig
 )
 
-func getHashForArtifact(ctx context.Context, depLister DependencyLister, a *latest.Artifact, mode config.RunMode) (string, error) {
+type artifactHasher interface {
+	hash(ctx context.Context, a *latest.Artifact) (string, error)
+}
+
+type Artifacts map[string]*latest.Artifact
+
+type artifactHasherImpl struct {
+	artifacts Artifacts
+	lister    DependencyLister
+	mode      config.RunMode
+	once      *util.Once
+}
+
+// newArtifactHasher returns a new instance of an artifactHasher. Use newArtifactHasherFunc instead of calling this function directly.
+func newArtifactHasher(artifacts Artifacts, lister DependencyLister, mode config.RunMode) artifactHasher {
+	return &artifactHasherImpl{
+		artifacts: artifacts,
+		lister:    lister,
+		mode:      mode,
+		once:      util.NewOnce(),
+	}
+}
+
+func (h *artifactHasherImpl) hash(ctx context.Context, a *latest.Artifact) (string, error) {
+	val := h.once.Do(func() interface{} {
+		return a.ImageName
+	},
+		func() interface{} {
+			hash, err := singleArtifactHash(ctx, h.lister, a, h.mode)
+			if err != nil {
+				return err
+			}
+			return hash
+		})
```

Comment:
so the scenario is that `singleArtifactHash` is getting called from multiple goroutines concurrently but I want to only run it once(any one goroutine) and have all the others wait until it completes and then all of them get the same result from the single processing. I think this is the simplest way of achieving this. 

Using a single `sync.Map` I can't block all goroutines from running, otherwise can you elaborate how it'd work with a single map?

Reply:
a little bit confused here. 
Is the `hash` calculated after the build seq is defined? ie. when is artifact cache look up done?

If its done after determining the build seq, then there would be only be 1 goroutine writing/computing the hash for a base artifact. Once it is computed and stored if say in a `sync.Map, multiple goroutines can read it at the same time. There is no need to block reads.

Reply:
It's in `BuildAndTest` function in `build_deploy.go`. We explicitly call `cache.Build` which computes the cache for all artifacts in parallel. For the filtered list of artifacts that need to be rebuilt we then call `InOrder` builder which orchestrates the builds in the correct order.

Reply:
So the cache computation isn't happening in the build pipeline per artifact but rather as a pre step.

Reply:
Pending
👍 This is great. Can I suggest a name change — a SyncStore? Why can't we pass in a key directly instead of a function?

Reply:
I'll keep `Once`. IMO from `Once.Do` it's clearer that it runs the function once. It's equivalent to `sync.Once.Do`. Also the `Do` method explains the usage so there shouldn't be any confusion.